### DISCUSSION TITLE
Troubleshooting strange bitrate loss during stream

### DISCUSSION THREAD

**Discussion by alexmck** [HaishinKit/HaishinKit.swift#1475] [Q&A]
Hi,

We are trying to troubleshoot and reproduce a sudden drop in bitrate for a very small subset of our users.

Most of our users will stream for around 60 minute using RTMP. During the middle of this, a small subset of users are seeing strange bitrate drops during the middle of their stream, before the bitrate recovers before the end of their stream.

Below are some sample graphs that illustrate this issue.

**User 1 (iPhone 8)**
![CleanShot 2024-06-13 at 12 31 40@2x](https://github.com/shogo4405/HaishinKit.swift/assets/4444107/3c02a8cf-7855-45b2-9226-9ace4c87d81f)

**User 2 (iPhone X)**
![CleanShot 2024-06-13 at 12 32 38@2x](https://github.com/shogo4405/HaishinKit.swift/assets/4444107/1382a557-91bc-48b9-818a-a2b330197c35)

**User 3 (iPad Air - not sure which generation)**
![CleanShot 2024-06-13 at 12 33 35@2x](https://github.com/shogo4405/HaishinKit.swift/assets/4444107/7b4deb82-eeb3-4754-969d-1eb864cbd139)

We have confirmed in each case these users have sufficient upload bandwidth, and I am confident internet speeds are not an issue. None of these users are using any adaptive bitrate settings.

User 2 also has an iPhone 13 Pro. When user 2 streams with this device, the problem goes away. User 2 also reports their iPhone X has battery health of 72% and will stream with the device plugged in.

User 3 also has a newer iPad Air. When users 3 streams with this device, the problem goes away.

All of these users are connecting to different streaming servers of ours in different regions, and there are other users streaming to these same servers at the same time without issue.

### What testing have we done?

We have tested both an iPhone 8 and an iPhone X and are unable to reproduce this issue.

We have also tested the iPhone 8 using Xcode and setting the device and setting Thermal State to both Serious and Critical, along with also testing GPU Performance State to minimum. Everything performed flawlessly during this time in our 3 hours of testing in these modes.

We have tested with devices while charging and while on battery.

We have tested using WiFi and using 4G/5G.

We have tested with phone cases on and off.

### When did this start to happen?

This problem started when we upgraded from either HaishinKit 1.5.6 or 1.5.7 to 1.7.2 or 1.7.3. We are currently using HaishinKit 1.8.1.

With the upgrade to HaishinKit 1.7 we switched to using HaishinKit's resampler. Before that we were using our own implementation.

These same users didn’t have this issue before this update to HaishinKit 1.7.

### What next?

I’m hoping for some more guidance as to what we could test next, or if there might have been a specific change in HaishinKit that could have caused this issue? I feel like we don’t have enough information yet to open an Issue for this problem.

**Comment by shogo4405** [comment]
I feel like there was a post yesterday... Was it deleted?

The behavior of the bitrate dropping is part of the ABR specification, so I don’t consider it an issue. Where do you see the problem?

**Comment by alexmck** [comment]
Zach deleted his post. It seems we are chasing 2 issues, that may or may not be related. One is this bitrate drop issue, and another is audio and video sync issues. Sometimes, the two issues seem to coincide.

I would like to note that ABR was turned off in these examples above.

**Comment by shogo4405** [comment]
`ABR was turned off` did you mean.

1. You are not changing the bitrate during the stream.
1. You are using videoCodecSettings.bitRateMode = .constantBitRate.

If you're using it in the sense of point 1, I think you should consider switching to .constantBitRate.

Additionally, from my experience, if there is a programming error where the PTS (Presentation Timestamp) of a video CMSampleBuffer is set to a past time just before sending, it can cause the bitrate to drop and take time to recover.

For example:
```
Frame 1 -> PTS = 1,000
Frame 2 -> PTS = 2,000
Frame 3 -> PTS = 3,000
Frame 4 -> PTS = 2,500
```

If you suspect that it’s related to audio sync issues, it could be due to the PTS values. There's also a chance this problem might occur even if the PTS is the same as the previous frame, but I haven’t tested this.

**Comment by zachsimone** [comment]
Hi @shogo4405 – yes, I posted a comment then deleted it. As Alex said I was conflating this issue with another, similar one we are having. Sorry. 

Thank you for the suggestion about the `bitRateMode` – we aren't setting it to `.constant` at the moment so I'll make that change in our project and see how it goes. If we're still having issues I'll look into the PTS. Thank you for that tip as well. :)
